File: DotEatGame/DotEatGame/Character.cpp
//======================================
//	ドットイートゲーム　Character
//======================================
#include "Character.h"
#include "Vector2.h"

// キャラクター初期化
void InitCharacter(Character* ch, Chara id, Vector2 initPos)
{
	//
	// ★ここをコーディングしてください
	// id,initPos で chのメンバーを初期化する
	// ch->lastPos は initPos を入れておく
	//
	ch->id = id;
	ch->pos = initPos;
	ch->lastPos = initPos;
}

// キャラクター移動
void MoveCharacter(Character* ch, Vector2 newPos)
{
	//
	// ★ここをコーディングしてください
	// 現在位置を lastPos に入れて
	// 現在位置に newPosをいれる
	//
	ch->lastPos = ch->pos;
	ch->pos = newPos;
}
// キャラクターの向きベクターを取得する
Vector2 GetCharacterDir(Character* ch)
{
	//
	// 単純に  pos-lastPos すると、ワープ地点でバグる
	//
	int x = ch->pos.x - ch->lastPos.x;
	if (x < -1 || 1 < x) {
		// 右端→左端、または 右端←左端
		x = (ch->pos.x == 0) ? +1 : -1;
	}
	int y = ch->pos.y - ch->lastPos.y;
	if (y < -1 || 1 < y) {
		// 下端→上端、または、下端←上端
		y = (ch->pos.y == 0) ? +1 : -1;
	}
	return { x,y };
}

File: DotEatGame_CS/DotEatGame_CS/Character.cs
//======================================
//	ドットイートゲーム　Character
//======================================
using Vector2 = GP2.Vector2;

namespace DotEatGame_CS
{
    enum Chara
    {
        Player, // プレーヤ
        Random, // きまぐれモンスター
        Chase,  // 追いかけモンスター
        Ambush, // 先回りモンスター
        Siege,  // 挟み撃ちモンスター
    }

    internal class Character
    {
        Chara m_id;        // 誰?
        Vector2 m_pos;     // 座標
        Vector2 m_lastPos; // 前回の座標
        Vector2 m_tmp;     // ワーク

        // プロパティ
        public Chara id
        {
            get { return m_id; }
        }
        public Vector2 pos
        {
            get { return m_pos; }
        }
        public Vector2 lastPos
        {
            get { return m_lastPos; }
        }
        // コンストラクタ
        public Character(Chara id, int initX, int initY)
        {
            m_id = id;
            m_pos.x = initX;
            m_pos.y = initY;
            m_lastPos = m_pos;
        }
        // キャラクター移動
        public void Move(Vector2 pos)
        {
            m_lastPos = m_pos;
            m_pos = pos;
        }
        // キャラクターの向きベクターを取得(ワープ対応)
        public Vector2 GetCharacterDir()
        {
            int x = m_pos.x - m_lastPos.x;
            if (x < -1 || 1 < x)
            {
                // 右端→左端、または 右端←左端
                x = (m_pos.x == 0) ? +1 : -1;
            }
            int y = m_pos.y - m_lastPos.y;
            if (y < -1 || 1 < y)
            {
                // 下端→上端、または、下端←上端
                y = (m_pos.y == 0) ? +1 : -1;
            }
            m_tmp.x = x;
            m_tmp.y = y;
            return m_tmp;
        }
    } // class
} // namespace